src/components/NewBranchDialog.tsx

```diff
@@ -0,0 +1,440 @@
+      <ListItem
+        button
+        className={classes(listItemClass, isBase ? activeListItemClass : null)}
+        key={idx}
+        onClick={this._onBranchClickFactory(branch.name)}
+      >
+        <span className={classes(listItemIconClass, 'jp-Icon-16')} />
+        <div className={listItemContentClass}>
+          <p
+            className={classes(
+              listItemTitleClass,
+              isBold ? listItemBoldTitleClass : null
+            )}
+          >
+            {branch.name}
+          </p>
+          {desc ? <p className={listItemDescClass}>{desc}</p> : null}
+        </div>
+      </ListItem>
+    );
+  }
+
+  /**
+   * Adds model listeners.
+   */
+  private _addListeners(): void {
+    // When the repository changes, we're likely to have a new set of branches:
+    this.props.model.repositoryChanged.connect(this._syncState, this);
+
+    // When the HEAD changes, decent probability that we've switched branches:
+    this.props.model.headChanged.connect(this._syncState, this);
+
+    // When the status changes, we may have checked out a new branch (e.g., via the command-line and not via the extension):
+    this.props.model.statusChanged.connect(this._syncState, this);
+  }
+
+  /**
+   * Removes model listeners.
+   */
+  private _removeListeners(): void {
+    this.props.model.repositoryChanged.disconnect(this._syncState, this);
+    this.props.model.headChanged.disconnect(this._syncState, this);
+    this.props.model.statusChanged.disconnect(this._syncState, this);
+  }
+
+  /**
+   * Syncs the component state with the underlying model.
+   */
+  private _syncState(): void {
+    const repo = this.props.model.pathRepository;
+    this.setState({
+      base: repo ? this.state.base : '',
+      current: repo ? this.props.model.currentBranch.name : '',
+      branches: repo ? this.props.model.branches : []
+    });
+  }
+
+  /**
+   * Callback invoked upon closing the dialog.
+   *
+   * @param event - event object
+   */
+  private _onClose = (): void => {
+    this.props.onClose();
+    this.setState({
+      name: '',
+      filter: ''
+    });
+  };
+
+  /**
+   * Callback invoked upon a change to the menu filter.
+   *
+   * @param event - event object
+   */
+  private _onFilterChange = (event: any): void => {
+    this.setState({
+      filter: event.target.value
+    });
+  };
+
+  /**
+   * Callback invoked to reset the menu filter.
+   */
+  private _resetFilter = (): void => {
+    this.setState({
+      filter: ''
+    });
+  };
+
+  /**
+   * Returns a callback which is invoked upon clicking a branch name.
+   *
+   * @param branch - branch name
+   * @returns callback
+   */
+  private _onBranchClickFactory(branch: string) {
+    const self = this;
+    return onClick;
+
+    /**
+     * Callback invoked upon clicking a branch name.
+     *
+     * @private
+     * @param event - event object
+     */
+    function onClick() {
+      self.setState({
+        base: branch
+      });
+    }
+  }
```

Comment:
Can be simplified as in the branch menu.

Reply:
Similar reply as [here](https://github.com/jupyterlab/jupyterlab-git/pull/518#discussion_r370911838).

Reply:
My motivation here was to use named functions rather than inline arrow functions for better debugging (e.g., via stack traces) and to elevate callback creation outside of rendering logic. 

Additionally, I find that, personally, when callbacks are [inlined](https://github.com/jupyterlab/jupyterlab-git/blob/576c4448f2ce19c63604a949e343da498cf81858/src/components/SinglePastCommitInfo.tsx#L230) within JSX, it convolves declarative rendering (this is what it should look like) with imperative programming logic (this is what it should do). Accordingly, my intuition is to elevate callback factories to the component prototype and use closures, with the added benefit of being able to provide JSDoc annotations for all functions.

Perhaps, as a matter of taste, this approach is a bit verbose, but IMO it strikes the right balance wrt separation of concerns and, in general, "debuggability".